File: utils/path_finding.py
from typing import Literal
from heapq import heappush, heappop

from langchain_core.runnables.config import P

from .base import get_rules, _parse_game_state, apply_move, get_state_positions, game_state_coords

def _blocked_entities(game_state: str, avoid_text: bool = True) -> list[list[int]]:
    """ Build simple "blocked" grid, where 1 means blocked (based on rules (e.g. wall is stop, lava is hot + you is melt)) and 0 means allowed. """
    rules = get_rules(game_state)
    _blocked_entities: list[tuple[str | None, str]] = [(None, "stop"), (None, "defeat"), ("melt", "hot")]
    yous = [r.entity for r in rules if r.state == "you"]
    matrix = _parse_game_state(game_state)
    blocked_matrix = []
    for y, row in enumerate(matrix):
        row_blocked = []
        for x, entity in enumerate(row):
            entities = entity.split("<")
            blocked = 0
            for e in entities:
                for you in yous:
                    for you_req, blocked_state in _blocked_entities:
                        rs = [r for r in rules if r.entity == e]
                        for r in rs:
                            if not you_req and r.state == blocked_state:
                                blocked = 1
                            elif you_req == you and r.state == blocked_state:
                                blocked = 1
                        if avoid_text and e.startswith("text_"):
                            blocked = 1
            row_blocked.append(blocked)
        blocked_matrix.append(row_blocked)
    return blocked_matrix


def heuristic(a: tuple[int,int], b: tuple[int,int]):
    return abs(a[0] - b[0]) + abs(a[1] - b[1])

def a_star(grid: list[list[int]], start: tuple[int,int], goal: tuple[int, int]) -> list[tuple[int,int]] | None:
    """Performs A* search on the grid. Returns the grid positions of the solution. Returns None if no solution has been found."""
    if grid[start[1]][start[0]] == 1 or grid[goal[1]][goal[0]] == 1:
            return None

    neighbors = [(0,1),(0,-1),(1,0),(-1,0)]
    shape = (len(grid[0]), len(grid))

    close_set: set[tuple[int, int]] = set()
    came_from: dict[tuple[int, int], tuple[int, int]] = {}
    gscore: dict[tuple[int, int], float] = {start:0}
    fscore: dict[tuple[int, int], float] = {start: heuristic(start, goal)}
    oheap = []

    heappush(oheap, (fscore[start], start))
    while oheap:
        current = heappop(oheap)[1]
        if current == goal:
            data: list[tuple[int,int]] = []
            while current in came_from:
                data.append(current)
                current = came_from[current]
            return data[::-1]
        close_set.add(current)
        for i, j in neighbors:
            neighbor = current[0] + i, current[1] + j
            tentative_g_score = gscore[current] + heuristic(current, neighbor)
            if 0 <= neighbor[0] < shape[0]:
                if 0 <= neighbor[1] < shape[1]:
                    if grid[neighbor[1]][neighbor[0]] == 1:
                        continue
                else:
                    # array bound y walls
                    continue
            else:
                # array bound x walls
                continue

            if neighbor in close_set and tentative_g_score >= gscore.get(neighbor, 0):
                continue

            if  tentative_g_score < gscore.get(neighbor, 0) or neighbor not in [i[1] for i in oheap]:
                came_from[neighbor] = current
                gscore[neighbor] = tentative_g_score
                fscore[neighbor] = tentative_g_score + heuristic(neighbor, goal)
                heappush(oheap, (fscore[neighbor], neighbor))

    return None

def convert_path_to_moves(path: list[tuple[int,int]]) -> list[Literal["up", "down", "left", "right"]]:
    lookup: dict[tuple, Literal["up", "down", "left", "right"]] = {
        (0,1): "down",
        (0,-1): "up",
        (1,0): "right",
        (-1,0): "left"
    }
    diff = lambda a, b: (a[0]-b[0], a[1]-b[1])
    diffs = [diff(b, a) for a, b in zip(path, path[1:])]
    return [lookup[p] for p in diffs]

# TODO: add tool to convert input to path and execute via
def shortest_path(game_state: str, goal: tuple[int, int], last_move: Literal["up", "down", "left", "right"]) -> list[Literal["up", "down", "left", "right"]]:
    # TODO: raise appropriate exceptions to pass reason for failure (e.g. goal is blocked, prev_goal is blocked etc.)
    # TODO: if goal is on text (e.g. to deactivate) it is handled not gracefully -> it should avoid text within the path but not if the goal is text
    goal = (goal[0] - 1, goal[1] - 1)
    goal_prev = apply_move(goal, last_move, reverse=True)
    blocked = _blocked_entities(game_state)
    blocked_no_text = _blocked_entities(game_state, avoid_text=False)
    if blocked_no_text[goal[1]][goal[0]]:
        return []
    prev_blocked = blocked[goal_prev[1]][goal_prev[0]]
    if prev_blocked:
        return []
    you_pos = get_state_positions(game_state, "you") # only take on "you" position

    for you in you_pos:
        you_coord = (you[0] - 1, you[1] - 1)

        path = a_star(blocked, you_coord, goal_prev)
        if path is None:
            continue
        moves = convert_path_to_moves([you_coord] + path)
        return moves + [last_move]
    return []


def reachable_entities(game_state: str) -> list[list[tuple[int, int, str]]]:
    coords = game_state_coords(game_state)
    reachable_entities: list[list[tuple[int, int, str]]] = []
    for row in coords:
        row_reachable = []
        for x, y, entity in row:
            path = None
            last_moves: list[Literal["up", "down", "left", "right"]] = ["up", "down", "left", "right"]
            for last_move in last_moves:
                p = shortest_path(game_state, (x, y), last_move)
                if len(p) > 0:
                    path = p
                    break
            if path:
                row_reachable.append((x, y, entity))
        if len(row_reachable) > 0:
            reachable_entities.append(row_reachable)
    return reachable_entities

File: .opencode/tools/path_finding.ts
import { type Direction, getRules, parseGameState, applyMove, getStatePositions, gameStateCoords } from "./base.js";
import type { Rule } from "./models.js";

interface PriorityQueueItem {
  priority: number;
  position: [number, number];
}

function pushHeap(heap: PriorityQueueItem[], item: PriorityQueueItem): void {
  heap.push(item);
  let i = heap.length - 1;
  while (i > 0) {
    const parent = Math.floor((i - 1) / 2);
    if (heap[parent]!.priority <= heap[i]!.priority) break;
    [heap[parent], heap[i]] = [heap[i]!, heap[parent]!];
    i = parent;
  }
}

function popHeap(heap: PriorityQueueItem[]): PriorityQueueItem | undefined {
  if (heap.length === 0) return undefined;
  const result = heap[0]!;
  const last = heap.pop()!;
  if (heap.length > 0) {
    heap[0] = last;
    let i = 0;
    while (true) {
      const left = 2 * i + 1;
      const right = 2 * i + 2;
      let smallest = i;
      if (left < heap.length && heap[left]!.priority < heap[smallest]!.priority) {
        smallest = left;
      }
      if (right < heap.length && heap[right]!.priority < heap[smallest]!.priority) {
        smallest = right;
      }
      if (smallest === i) break;
      [heap[i], heap[smallest]] = [heap[smallest]!, heap[i]!];
      i = smallest;
    }
  }
  return result;
}

function blockedEntities(gameState: string, avoidText: boolean = true): number[][] {
  const rules = getRules(gameState);
  const blockedStates: [string | null, string][] = [[null, "stop"], [null, "defeat"], ["melt", "hot"]];
  const yous = rules.filter((r) => r.state === "you").map((r) => r.entity);
  const matrix = parseGameState(gameState);

  const blockedMatrix: number[][] = [];

  for (let y = 0; y < matrix.length; y++) {
    const rowBlocked: number[] = [];
    for (let x = 0; x < matrix[y]!.length; x++) {
      const entity = matrix[y]![x]!;
      const entities = entity.split("<");
      let blocked = 0;

      for (const e of entities) {
        for (const you of yous) {
          for (const [youReq, blockedState] of blockedStates) {
            const matchingRules = rules.filter((r) => r.entity === e);
            for (const r of matchingRules) {
              if (!youReq && r.state === blockedState) {
                blocked = 1;
              } else if (youReq === you && r.state === blockedState) {
                blocked = 1;
              }
            }
          }
        }
        if (avoidText && e.startsWith("text_")) {
          blocked = 1;
        }
      }
      rowBlocked.push(blocked);
    }
    blockedMatrix.push(rowBlocked);
  }
  return blockedMatrix;
}

function heuristic(a: [number, number], b: [number, number]): number {
  return Math.abs(a[0] - b[0]) + Math.abs(a[1] - b[1]);
}

type Node = [number, number];

function aStar(
  grid: number[][],
  start: Node,
  goal: Node
): Node[] | null {
  if (
    grid[start[1]]?.[start[0]] === 1 ||
    grid[goal[1]]?.[goal[0]] === 1
  ) {
    return null;
  }

  const neighbors: [number, number][] = [
    [0, 1],
    [0, -1],
    [1, 0],
    [-1, 0],
  ];
  const shape: [number, number] = [grid[0]!.length, grid.length];

  const closeSet = new Set<string>();
  const cameFrom = new Map<string, Node>();
  const gScore = new Map<string, number>();
  const fScore = new Map<string, number>();
  const oheap: PriorityQueueItem[] = [];

  const startKey = `${start[0]},${start[1]}`;
  gScore.set(startKey, 0);
  fScore.set(startKey, heuristic(start, goal));
  pushHeap(oheap, { priority: heuristic(start, goal), position: start });

  while (oheap.length > 0) {
    const current = popHeap(oheap)!.position;
    const currentKey = `${current[0]},${current[1]}`;

    if (current[0] === goal[0] && current[1] === goal[1]) {
      const path: Node[] = [];
      let curr: Node | undefined = current;
      while (curr) {
        path.unshift(curr);
        curr = cameFrom.get(`${curr[0]},${curr[1]}`);
      }
      return path;
    }

    closeSet.add(currentKey);

    for (const [di, dj] of neighbors) {
      const neighbor: Node = [current[0] + di, current[1] + dj];
      const neighborKey = `${neighbor[0]},${neighbor[1]}`;

      if (
        0 <= neighbor[0] &&
        neighbor[0] < shape[0] &&
        0 <= neighbor[1] &&
        neighbor[1] < shape[1]
      ) {
        if (grid[neighbor[1]][neighbor[0]] === 1) {
          continue;
        }
      } else {
        continue;
      }

      if (closeSet.has(neighborKey)) continue;

      const tentativeGScore =
        (gScore.get(currentKey) ?? Infinity) + heuristic(current, neighbor);

      if (
        tentativeGScore < (gScore.get(neighborKey) ?? Infinity) ||
        !oheap.some((item) => item.position[0] === neighbor[0] && item.position[1] === neighbor[1])
      ) {
        cameFrom.set(neighborKey, current);
        gScore.set(neighborKey, tentativeGScore);
        fScore.set(neighborKey, tentativeGScore + heuristic(neighbor, goal));
        pushHeap(oheap, {
          priority: tentativeGScore + heuristic(neighbor, goal),
          position: neighbor,
        });
      }
    }
  }

  return null;
}

const directionLookup: Record<string, Direction> = {
  "0,1": "down",
  "0,-1": "up",
  "1,0": "right",
  "-1,0": "left",
};

export function convertPathToMoves(path: Node[]): Direction[] {
  const diffs: [number, number][] = [];
  for (let i = 0; i < path.length - 1; i++) {
    const a = path[i]!;
    const b = path[i + 1]!;
    diffs.push([b[0] - a[0], b[1] - a[1]]);
  }
  return diffs.map((d) => directionLookup[`${d[0]},${d[1]}`] as Direction);
}

export function shortestPath(
  gameState: string,
  goal: [number, number],
  lastMove: Direction
): Direction[] {
  const goalCoord: [number, number] = [goal[0] - 1, goal[1] - 1];
  const goalPrev = applyMove(goalCoord, lastMove, true);
  const blocked = blockedEntities(gameState);
  const blockedNoText = blockedEntities(gameState, false);

  if (blockedNoText[goalCoord[1]]?.[goalCoord[0]]) {
    return [];
  }

  if (blocked[goalPrev[1]]?.[goalPrev[0]]) {
    return [];
  }

  const youPositions = getStatePositions(gameState, "you");

  for (const you of youPositions) {
    const youCoord: [number, number] = [you[0] - 1, you[1] - 1];
    const path = aStar(blocked, youCoord, goalPrev);

    if (path === null) {
      continue;
    }

    const moves = convertPathToMoves([youCoord, ...path]);
    return [...moves, lastMove];
  }

  return [];
}

export function reachableEntities(
  gameState: string
): [number, number, string][] {
  const coords = gameStateCoords(gameState);
  const reachable: [number, number, string][] = [];

  for (const [x, y, entity] of coords) {
    const lastMoves: Direction[] = ["up", "down", "left", "right"];
    let pathFound = false;

    for (const lastMove of lastMoves) {
      const p = shortestPath(gameState, [x, y], lastMove);
      if (p.length > 0) {
        pathFound = true;
        break;
      }
    }

    if (pathFound) {
      reachable.push([x, y, entity]);
    }
  }

  return reachable;
}
